### PR TITLE
[Feature] Archive and un-archive pools

### LINKED ISSUE #6578: ✨ Enable soft delete to archive pools
## ✨ Feature

We have had this button (non-functional) on the "Edit pool" page for a long time and it's becoming more and more useful as pools start to expire. This would remove the clutter from the current Pools table and make them easier to manage.

### "Archive" button

The "Archive" button should only be visible if the pool's current status is "Closed". 

The "Archive" button should prompt a dialog to confirm the archive action.

When confirmed, the pool's status should change from "Closed" to "Archived"

### "Un-archive" button 

When a pool has been archived. A new button appears  next to this status to "Un-archive" this pool.

The "Un-archive" button should only be visible if the pool's current status is "Archived"

The "Un-archive" button should prompt a dialog to confirm the un-archive action.

When confirmed, the pool's status should change from "Archived" to "Closed"

### ⚙️ New functionality in other pages

"Archived" pools should be **hidden by default** in the Pools table

"Archived" pools should **NOT show up** on a the request page as options to submit requests. Nor it's candidates be counted towards the request results.

"Archived" pools should **NOT show up** on lists of pools available for "Add user to pool" in the pool information section on a user's information page.

"Archived" pools should **NOT show up** on lists of pools available for "Batch change status" in the pool information section on a user's information page.

On the candidate public facing side, an "Archived" pool behaves the same as a "Closed" pool for the purposes of any logic tied to it.

## 🎨 Design File

Include a [link to the design file](https://www.figma.com/file/tH8X3HEgXsQVGTa2eheH8P/Pool-Updates---Admin-(elisa)?type=design&node-id=1618%3A6064&t=CQyQmZR1UhCudnQW-1) 

## 📸 Screenshot

image

image

image

image

## 🙋‍♀️ Proposed Implementation

Use the soft-delete functionality built-in to Eloquent models.  The deleted_at column already exists in the database table.

### PR DESCRIPTION
**Pull request by petertgiles**
🤖 Resolves #6578 

## 👋 Introduction

This branch adds the ability to archive and un-archive pools.

## 🕵️ Details

The functionality is based on a new timestamp `archived_at` and associated scope.  Based on [this comment](https://talent-cloud.slack.com/archives/C0259G6KXJ6/p1689168553019879?thread_ts=1689106800.319769&cid=C0259G6KXJ6), a new permission was added to the matrix to check if the user has the ability to archive a pool.  Based on [this comment](https://talent-cloud.slack.com/archives/C0259G6KXJ6/p1689692066451029?thread_ts=1689691611.551129&cid=C0259G6KXJ6), the same permission was used to check when un-archiving the archived pool.

Based on [this comment](https://talent-cloud.slack.com/archives/C0259G6KXJ6/p1689168580183789?thread_ts=1689106800.319769&cid=C0259G6KXJ6), an issue (#7306) was opened to convert the draft-delete operation to a hard-delete.

Based on the issue AC and following discussion, archived pools do not show up in the UI in any obvious way.  It sounds like an issue will come later to update the pool table and filters.  In the mean time, you can bring up the pool page for an archived pool either by putting the UUID in the URL manually or finding it through an existing pool candidate.  Archived pools can still be found through relationships or when queried by specific ID.

## 🧪 Testing

> __Note__
I found testing this was easiest by setting up the environment with the CI seeder and manually creating the pools to test with.

1. Rerun the roles and permissions seeder `./artisan db:seed --class=RolePermissionSeeder`
2. Rerun the "Refresh API" task
3. Rebuild the frontend
4. Log into the admin site and navigate to `/en/admin/pools`
5. Work through the AC on the issue

## 📸 Screenshot

image

# :truck: Deployment instructions
- Rerun the roles and permissions seeder to get the new rows.

### CODE REVIEW

File: api/app/GraphQL/Mutations/ArchivePool.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    /**
+     * Closes the pool by setting the archived_at to now().
+     * @param  null  $_
+     * @param  array{}  $args
+     */
+    public function __invoke($_, array $args)
+    {
+        $pool = Pool::find($args['id']);
+        if ($pool->getStatusAttribute() !== ApiEnums::POOL_IS_CLOSED) {
+            throw new Error("You cannot archive a pool unless it is in the closed status.");
```
Comment (petertgiles) [inline api/app/GraphQL/Mutations/ArchivePool.php:21]: Does anyone know how get an error to show up in the GraphQL response?  I think this will show up as a generic HTTP 500 error.  Not a huge deal since the UI will prevent it from happening but it seems like there should be a better way.

File: api/graphql/schema.graphql [hunk 2]
```diff
@@ -878,17 +879,16 @@ type Query {
   # countApplicants returns the number of candidates matching its filters, and requires no special permissions.
   countApplicants(where: ApplicantFilterInput): Int! @count(model: "User")
   pool(id: UUID! @eq): Pool @find @can(ability: "view", query: true)
-  pools: [Pool]! @all @can(ability: "viewAny", model: "Pool")
```
Comment (petertgiles) [inline api/graphql/schema.graphql:881]: Very strange - there was a duplicate query in the schema.  It seems like the last definition in the file won so I removed this one.

File: apps/web/src/pages/Teams/TeamMembersPage/components/RemoveTeamMemberDialog.tsx [hunk 3]
```diff
@@ -123,12 +127,7 @@ const RemoveTeamMemberDialog = ({
             </>
           ) : null}
           <p data-h2-margin="base(x1, 0)">
-            {intl.formatMessage({
-              defaultMessage: "Do you wish to continue?",
-              id: "2JIgrl",
-              description:
-                "Message displayed to users before removing a member from a team",
-            })}
+            {intl.formatMessage(uiMessages.confirmContinue)}
```
Comment (petertgiles) [inline apps/web/src/pages/Teams/TeamMembersPage/components/RemoveTeamMemberDialog.tsx:130]: I decided to centralize this message.

**Review by petertgiles — COMMENTED** [review]
A few comments...

File: api/app/GraphQL/Mutations/ArchivePool.php (hunk 1 above)
Comment (vd1992) [reply api/app/GraphQL/Mutations/ArchivePool.php:21]: Do you mean something like below?

image

image

**Comment by vd1992** [comment]
You've added two mutations, a scope, and a new policy. I think some unit tests would be a good addition. I don't think you need to test every little change to backend, but I think the mutations ought to be tested at the minimum

File: api/app/Policies/PoolPolicy.php [hunk 4]
```diff
@@ -192,4 +192,17 @@ public function deleteDraft(User $user, Pool $pool)
...
+     *
+     * @param  \App\Models\User  $user
+     * @param  \App\Models\Pool  $pool
+     * @return \Illuminate\Auth\Access\Response|bool
+     */
+    public function archive(User $user, Pool $pool)
+    {
+        $pool->loadMissing('team');
+        return $user->isAbleTo("archive-team-pool", $pool->team);
+    }
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Policies/PoolPolicy.php:207]: This is going to sound really pedantic on my part, but I'm iffy on the use of the word "restore" to refer to un-archiving something. Restore in Laravel tends to be used pretty commonly with undoing deletion, so is probably a word we should be careful with to avoid confusing ourselves down the line 😅

File: api/app/GraphQL/Mutations/ArchivePool.php (hunk 1 above)
Comment (petertgiles) [reply api/app/GraphQL/Mutations/ArchivePool.php]: Yes, that is exactly what I needed!  Thanks.
Ref: [validation errors](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/39e7c859e160799abe3bac0816476670dbc6f6a5)

File: api/app/Policies/PoolPolicy.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (petertgiles) [reply api/app/Policies/PoolPolicy.php:207]: >  I'm iffy on the use of the word "restore" 

Yeah, that's fair.  I originally used soft-deletes where it made more sense but not as much now.  Un-archive is super clunky but it matches the frontend so I'll go with that.
Ref: [restore to unarchive](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/50621bfbb81e9f99106a97a274ef3386f7f468ff)

[the remark this answers is not in the document]
Comment (petertgiles) [reply api/app/Policies/PoolPolicy.php]: > restore also uses this policy

Updated the policy name and comment.
Ref: [policy name](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/9d5128b04355a1baee9e8ee8efb530c7ab05c2ef)

**Comment by petertgiles** [comment]
> I feel some unit tests would be a good addition.

You're right!  :sweat_smile: 
Ref: [tests](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/a8715ad8d522c44e1d29bb39b4cc902b877e2a06)

File: api/app/Models/Pool.php [hunk 5]
```diff
@@ -122,29 +124,19 @@ public function screeningQuestions(): HasMany
     public function getStatusAttribute()
     {
         // given database is functioning in UTC, all backend should consistently enforce the same timezone
```
Comment (vd1992) [inline api/app/Models/Pool.php:126]: This comment is from before, but I don't think it is needed for the current state of the accessor and makes it slightly more confusing. Please delete it.

File: api/graphql/schema.graphql
Comment (vd1992) [inline api/graphql/schema.graphql:1]: If you log in as a pool operator, you can still see archived pools
Pool operators see pools differently, it is programmatic-ly determined from their `roleAssignments`

image

image

File: api/app/Policies/PoolPolicy.php
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply api/app/Policies/PoolPolicy.php:1]: @vd1992 I think the "batch change" is referring to this dropdown: 
image

File: api/app/Models/Pool.php [hunk 6]
```diff
@@ -189,4 +181,13 @@ public function scopeAuthorizedToView(Builder $query)
 
         return $query;
     }
+
+    public function scopeNotArchived(Builder $query, ?array $args)
```
Comment (tristan-orourke) [inline api/app/Models/Pool.php:185]: You can leave out the $args prop since it's not being used.

File: api/tests/Feature/PoolTest.php [hunk 7]
```diff
@@ -204,6 +241,8 @@ public function testPoolQueryReturnsOnlyPublished(): void
         $draftPool = Pool::factory()->create([
             'published_at' => null,
         ]);
+        // this pool is archived so it should not be returned in the publishedPool query
```
Comment (tristan-orourke) [inline api/tests/Feature/PoolTest.php:244]: You can leave this here, but I'd recommend making a separate test to check that archived pools aren't returned in the main query. Having an atomic test like that would make it more obvious what's happening if it failed.

File: api/app/Policies/PoolPolicy.php
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply api/app/Policies/PoolPolicy.php:1]: We don't expect there to be many pool candidates in archived pools, but if there are, I think it would be a good idea to keep showing them here.

**Review by tristan-orourke — APPROVED** [review]
Looks pretty good, and @vd1992 has already covered a lot. 

I just want to bring attention to this acceptance criteria:
>"Archived" pools should NOT show up on a the request page as options to submit requests. **Nor it's candidates be counted towards the request results**.

Technically, candidates in archived pools won't show up in any counts right now, because we're first querying for pools (which excludes archived) and then including all those pools in the countApplicants queries. However, I suspect this might change when we refactor the page , since filtering by pools is _mostly_ redundant with filtering by classification and stream. And if you don't filter by pools explicitely, countApplicants will happily return candidates from archived pools.

UPDATE - I was going to suggest how to scope archived candidates out of the count queries, but then I remembered the decision @gobyrne and I made that archiving a pool _shouldn't_ "soft-delete" any candidates. So the current behaviour is fine. It's likely a moot point anyway, since in practice pools are unlikely to be archived while they have available candidates.

File: api/graphql/schema.graphql
Comment (tristan-orourke) [reply api/graphql/schema.graphql:1]: Good catch! I'll pull https://github.com/GCTC-NTGC/gc-digital-talent/issues/6109 in as an interruption, since our admin team will be very confused by this.

File: api/app/Policies/PoolPolicy.php
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Policies/PoolPolicy.php:1]: Aye aye

File: api/app/GraphQL/Mutations/UnarchivePool.php [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+namespace App\GraphQL\Mutations;
+
+use App\Models\Pool;
+use Database\Helpers\ApiEnums;
+use Nuwave\Lighthouse\Exceptions\ValidationException;
+
+final class UnarchivePool
+{
+    /**
+     * Closes the pool by setting the archived_at to now().
```
[the remark this answers is not in the document]
Comment (petertgiles) [reply api/app/GraphQL/Mutations/UnarchivePool.php]: :+1: 
[fix copied comment](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/c58c51990a5182942a44e2bce889ca8b3b9a7ab7)

File: api/app/Models/Pool.php (hunk 6 above)
Comment (petertgiles) [reply api/app/Models/Pool.php]: :+1: 
[Remove unused parameter](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/97984ab44ddf354bf092c781082d2017b2622bbb)

File: api/app/Models/Pool.php (hunk 5 above)
Comment (petertgiles) [reply api/app/Models/Pool.php]: :+1: 
[remove old comment](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/a8fde02611df2a8b462951178d6c4329ffc13e7b)

File: api/app/Models/Pool.php
[the remark this answers is not in the document]
Comment (petertgiles) [reply api/app/Models/Pool.php:1]: :+1: 
[add missing cast](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/baac25b8171f58b5679bfcfb906baa37680000ec)

File: api/tests/Feature/PoolTest.php (hunk 7 above)
Comment (petertgiles) [reply api/tests/Feature/PoolTest.php]: Yeah, smaller units tests are better.  I've just split the test in half.
[split test into two](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/9959bc4c71920a2a339dd4845054caabd381b43d)

Comment (petertgiles) [reply api/tests/Feature/PoolTest.php]: And [split anon test](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7399/commits/a62be9da70456ff762b6d6f670102a59f45a5743)
